### ISSUE TITLE
Loop order affects speed x10 when it shouldn't

### ISSUE THREAD

**Issue by nidleo** [numba/numba#1241]
I have the following code which simulates extracting binary words from different locations within a set of images.

The Numba wrapped function, wordcalc in the code below, has 2 problems:
1. It is 3 times slower compared to a similar implementation in C++.
2. Most strangely, if you switch the order of the "ibase" and "ibit" for-loops, speed drops by a factor of 10!. This does not happen in the C++ implementation which remains unaffected.

Edit: I'm using Numba 0.18.2 from WinPython 2.7

The code:

    #%%
    import numpy as np

    imDim = 80
    numInsts = 10**4
    numInstsSub = 10**4/4
    bitsNum = 13;

    Xs = np.random.rand(numInsts, imDim**2)       
    iInstInds = np.array(range(numInsts)[::4])
    baseInds = np.arange(imDim**2 - imDim*20 + 1)
    ofst1 = np.random.randint(0, imDim*20, bitsNum)
    ofst2 = np.random.randint(0, imDim*20, bitsNum)

    @nb.jit(nopython=True)
    def wordcalc(Xs, iInstInds, baseInds, ofst, bitsNum, newXz):
        count = 0
        for i in iInstInds:
            Xi = Xs[i]        
            for ibase in range(baseInds.shape[0]):    
                for ibit in range(bitsNum):
                    u = Xi[baseInds[ibase] + ofst[0, ibit]] > Xi[baseInds[ibase] + ofst[1, ibit]]
                    newXz[count, ibase] = newXz[count, ibase] | np.uint16(u * (2**ibit))
            count += 1
        return newXz
                
    ret = wordcalc(Xs, iInstInds, baseInds, np.array([ofst1, ofst2]), bitsNum, np.zeros((iInstInds.size,     baseInds.size), dtype=np.uint16))

**Comment by seibert** [comment]
I'm not sure why the 3x performance difference relative to C++, but with Numba 0.19.1 I only see a 40% difference in performance when I swap the order of the ibase and ibit for loops:
```python
%timeit wordcalc(Xs, iInstInds, baseInds, np.array([ofst1, ofst2]), bitsNum, np.zeros((iInstInds.size,     baseInds.size), dtype=np.uint16))
%timeit wordcalc_flip(Xs, iInstInds, baseInds, np.array([ofst1, ofst2]), bitsNum, np.zeros((iInstInds.size,     baseInds.size), dtype=np.uint16))
1 loops, best of 3: 1.97 s per loop
1 loops, best of 3: 1.18 s per loop
```
The size of this difference seems more in line with what I would expect from cache effects, though I haven't analyzed exactly why the flipped loop would be faster.

**Comment by seibert** [comment]
OK, I think I can guess why the C++ code is 3x faster.  Looking at the generated assembly, it seems that the LLVM compiler is not recognizing this expression:
```
u * (2**ibit)
```
is equivalent to the much faster:
```
u << ibit
```
If I make that change to the Python code manually, the original wordcalc runs in 0.503 seconds, and the flipped version runs in 0.487 seconds.

**Comment by nidleo** [comment]
Thanks!
This change really solves the strange loop-order behavior and brings the speed closer to C++.
Now Numba runs in 0.74 seconds while C++ runs in 0.43 seconds. Any ideas why the gap exists? Could upgrading to 0.19 solve this?

One more thing that could also be related to the speed gap: changing to the faster
    u << ibit
did not work at first since "u" was treated as int8 and when ibit>8, the operation resulted in 0.
I then had to use
    np.uint16(u+0) << ibit
The +0 is necessary since otherwise I'd get: "TypingError: Undeclared Function(<type 'numpy.uint16'>)(bool,)"
Is there any better way of making this work?

**Comment by seibert** [comment]
Hmm.  That sounds like a missing type signature for the cast function.  We should add that.

**Comment by pitrou** [comment]
> The +0 is necessary since otherwise I'd get: "TypingError: Undeclared Function()(bool,)"

It seems this is already fixed in Numba master (and probably 0.20).

The fact that ``1 << bits`` can lose some information because 1 is typed as int8 is another issue that would probably deserve addressing, I think.

**Comment by nidleo** [comment]
Thanks for following up on this. I upgraded to 0.20 and the "+0" problem was indeed fixed.

I'm closing this issue as most of the problems were addressed, but I think some thought should be put into the int8 conversion. Different conversion rules make the python and numba versions inconsistent. Not only that, this inconsistency is somewhat hard to pick up on - the difference is exhibited only for large enough bitsNum.
It would be safer to throw an error that requests explicit conversion if automatic conversion is not able to guarantee consistency.
